### ISSUE TITLE
Velocity smoothing MINRES takes 3497 iterations and is 75% of a relaxation step

### ISSUE THREAD

**Issue by alankaptanogluconsulting** [ToBlick/mrx#18]
## Summary

One `relaxation_step` at li383 `(12,24,12)` p=3 float32 performs **37 478 operator applies**. Three MINRES solves are 99.5% of them, and the velocity smoothing solve is 75% on its own, at **3 497 iterations**. The six inverse-mass CG solves normally assumed to be the cost take 20-27 iterations and are 0.5%.

This is backend-independent. It surfaced while benchmarking a Cloud TPU v5e (#17), where one apply costs ~0.35 ms and the step therefore takes 13 s, but the same iteration counts hold on CPU and GPU, where each apply is roughly 30x cheaper and the cost is easier to miss.

## Measurement

Nothing reads these counts today because `relaxation_loop` is `@jax.jit` around a `lax.scan`, so every solver's `info` is discarded before it becomes concrete. Running one step eagerly makes it readable. Default production `TimeStepper` (LBFGS, history 1, analytic line search, `velocity_smoothing_order=1`, `scale = 0.064 / n_r^2`), `tol = sqrt_eps = 3.45e-04`, `maxiter = 10000`:

| # | solver | n | iterations | converged | applies |
|---|---|---|---|---|---|
| 0 | CG, inverse mass | 8700 | 0 | yes | 0 |
| 1 | CG, inverse mass k=1 | 8124 | 25 | yes | 50 |
| 2 | CG, inverse mass | 8376 | 0 | yes | 0 |
| 3 | MINRES, Leray k=3 | 2880 | 221 | yes | 3 536 |
| 4 | CG, inverse mass k=2 | 8376 | 27 | yes | 54 |
| 5 | **MINRES, velocity smoothing k=2** | 8376 | **3 497** | yes | **27 976** |
| 6 | MINRES, Leray k=3 | 2880 | 361 | yes | 5 776 |
| 7 | CG, inverse mass k=2 | 8376 | 22 | yes | 44 |
| 8 | CG, inverse mass k=1 | 8124 | 21 | yes | 42 |

Solve 5 is `apply_inverse_mass_plus_eps_laplace_matrix` on `(M_2 + eps L_2) x = M_2 u`, reached through `TimeStepper.smooth_velocity` and solved by `solve_saddle_point_minres`.

## Why this looks like preconditioning rather than precision

- **It scales linearly with the DoFs.** At `(8,16,8)`, `n(2) = 2192`, solve 5 takes 951 iterations; at `(12,24,12)`, `n(2) = 8376`, it takes 3 497. DoFs up 3.8x, iterations up 3.7x. A preconditioner that controlled the condition number would give a roughly flat count.
- **Refining should make it easier, not harder.** `eps = 0.064 / n_r^2` shrinks as you refine, so `(M_2 + eps L_2)` becomes *more* mass-dominated at higher resolution while the iteration count rises anyway.
- **It is not stagnation and not float32.** Every solve converges and returns a negative `info`; none hits `maxiter`. Inverse-mass CG at the same tolerance takes the same iteration count in float32 on a TPU MXU as on a host CPU (20 at k=1, 24 at k=2), so the arithmetic precision is not the issue.

The upper block currently takes the mass-dominated (metric-lumping mass) preconditioner. That choice looks right for small `eps` in principle, and the measurement says it is not delivering.

## Why it matters even though smoothing is optional

`velocity_smoothing_order=1` is the tutorial and production setting, and it earns its cost: the docstring notes li383 reaches a clean nested floor in ~1000 steps with it against ~6000 without. But it makes the step ~4x more expensive, so the net gain over the unsmoothed descent is only about 1.5x. Fixing the iteration count would convert most of that 4x into real speedup, on every backend.

## Reproducing

Wrap the two entry points `mrx/operators.py` binds, `solve_singular_cg` and `solve_saddle_point_minres`, record the returned signed `info`, then call `ts.relaxation_step(initial_state(B0, ts, 1.0))` eagerly rather than through `relaxation_loop`. Full context and the TPU-side numbers are in `tpu/results/benchmark_v5e_vs_cpu.md` on the `tpu-support` branch.

**Comment by akaptano** [comment]
## Measured, and the cause is a wrong comment rather than a missing preconditioner

The suspicion above was that the mass preconditioner was the problem but "that choice looks right for small `eps` in principle". It is not right, and the reason it looked right is a claim in the code that had never been checked.

```
mrx/operators.py
# M + eps L is MASS-DOMINATED in the regime this solve is used in
# (the relaxation's velocity smoothing: eps * lambda_max(M^-1 L) ~ 0.26
# at ns=(8,16,8)), so the production MASS preconditioner is the right
```

Power iteration on `M^-1 L_hodge`, the operator this solve actually contains:

| `ns` | `eps` | `lambda_max(M^-1 L)` | `eps * lambda_max` |
|---|---|---|---|
| (8,16,8) | 1.00e-03 | 9.15e+04 | **91.5** |
| (12,24,12) | 4.44e-04 | 8.11e+05 | **360.3** |

Two things follow. The claimed 0.26 is wrong by 352x, so `M + eps L` is strongly Laplacian-dominated and the mass atom is preconditioning the term that does not dominate. And it is not resolution-flat: `lambda_max` grows 8.9x between these two resolutions while `eps` falls only 2.25x, so refining makes it worse, which is exactly the 951 -> 3 497 iteration growth reported above. The MINRES residual history falls steadily rather than plateauing, so this is conditioning and not stagnation.

## What was tried

At both resolutions, iterations for the smoothing solve:

| upper-block preconditioner | (8,16,8) | (12,24,12) | growth |
|---|---|---|---|
| mass, `P_M` (shipping) | 2 130 | 8 493 | 4.0x |
| **Laplacian, `(1/eps) P_L`** | **1 655** | **3 636** | **2.2x** |
| `jacobi` | 14 372 | diverges | - |
| none | 21 022 | diverges | - |

The Laplacian atom wins on both count and scaling. `P_M + (1/eps) P_L` and a Neumann-series preconditioner were also tried and were worse than `(1/eps) P_L` alone. Preconditioned CG on the reduced SPD operator `M + eps L`, with `L` applied through an inner mass solve, cuts outer iterations 18x but spends all of it in the inner solve (54 264 inner CG iterations) and lands at wall-clock parity; worth recording that an inexact inner solve costs only 3% more outer iterations, so flexible CG is not the obstacle -- the inner solve simply is not cheap. Sweeping `eps` puts the crossover where the mass atom would be the right choice at `eps ~ 1e-4`, below the regime this solve uses.

## Fix

#19 adds a `laplacian` kind and plumbs it through `TimeStepper.smooth_velocity`, which previously could not pass a preconditioner at all. On a v5e at `(12,24,12)` p=3 float32 the whole relaxation step goes **13.02 s to 8.05 s, a 1.62x**, timed against the old preconditioner on the same node, with the trajectory unchanged to 4.3e-09 against a 1.49e-08 solver tolerance. `test/test_diffusion_preconditioner.py` pins the ordering so it cannot drift again unnoticed.

This is a large constant, not a cure: the condition number is still resolution-dependent, and the growth is 2.2x rather than 4.0x per refinement but is not flat. The object that would make it flat is the separable `(M + eps L)` atom that `docs/research/OPEN.md` section 3.9 already proposes for the sibling resistive solve, which reports the same symptom at 383 iterations/step for `eta=1e-3`. Both solves would be fixed by the same atom, so I would keep this issue open against that.

**Comment by ToBlick** [comment]
Solvers are improved (a000438). At the end of the day, inverting a vector-Laplace problem on these strongly-shaped geometries with non-orthogonal mappings is just expensive. A few thousand iterations are not all that unexpected. See the response to #19 for more on the shifted solves. Closing this for now, re-open if it did not improve substantially.
